File: apps/gui/src/lib/model_alias.ts
/**
 * 模型简称归一 —— TS 镜像 of `kode-core/src/model_alias.rs`。
 *
 * 两边实现必须等价,详见 `tests/model_alias_fixtures.json` 共享夹具。
 *
 * 输入示例:
 *   "Claude-Sonnet-4.6 (1M context)" → "sonnet-4.6-1m"
 *   "claude-opus-4.7-1m"             → "opus-4.7-1m"
 *   "anthropic.claude-haiku-4.5"     → "haiku-4.5"
 *   "gpt-5.3-codex"                  → "gpt-5.3-codex"
 */

export function shortModelName(raw: string): string {
  const cleaned = sanitizeModelName(raw)
  if (!cleaned || cleaned === 'auto') return cleaned

  // 1) 拆括号
  const { body, parenSuffix } = splitParen(cleaned)

  // 2) 主体小写 + 空格→dash
  const lowered = body.trim().replace(/\s+/g, '-').toLowerCase()

  // 3) 去前缀
  const stripped = lowered.replace(/^anthropic\./, '').replace(/^claude-/, '')

  // 4) 数字段合并 + 丢日期
  const parts = stripped.split('-').filter((x) => x.length > 0)
  // 4.5) claude code 用 "claude-<ver>-<tier>";统一成 "<tier>-<ver>" 再 compact
  const swapped = swapVerTierIfNeeded(parts)
  const compact = compactParts(swapped)

  // 5) 拼上括号 size 标签
  if (parenSuffix && !compact.includes(parenSuffix)) {
    return `${compact}-${parenSuffix}`
  }
  return compact
}

export function sanitizeModelName(raw: string): string {
  const s = (raw ?? '').trim()
  if (!s) return ''

  const lower = s.toLowerCase()
  let cut = s.length
  for (const marker of [
    ' note:',
    '-note:',
    '\nnote:',
    '\r\nnote:',
    '\tnote:',
    ' the model was saved to user settings',
    '-the-model-was-saved-to-user-settings',
  ]) {
    const idx = lower.indexOf(marker)
    if (idx >= 0) {
      cut = Math.min(cut, idx)
    }
  }

  // 兜底:合法 model 名不含 \n / \r / \t(友好名最多用空格)。
  // 这三个字符出现一定是 codebuddy 又夹带了新的 note 形态。
  for (let i = 0; i < s.length; i++) {
    const c = s.charCodeAt(i)
    if (c === 10 || c === 13 || c === 9) {
      cut = Math.min(cut, i)
      break
    }
  }

  return s.slice(0, cut).replace(/[ -]+$/, '').trim()
}

function splitParen(s: string): { body: string; parenSuffix: string | null } {
  const lp = s.indexOf('(')
  const rp = s.lastIndexOf(')')
  if (lp >= 0 && rp > lp) {
    const body = (s.slice(0, lp) + s.slice(rp + 1)).trim()
    const inside = s.slice(lp + 1, rp)
    return { body, parenSuffix: extractSizeTag(inside) }
  }
  return { body: s, parenSuffix: null }
}

function extractSizeTag(inside: string): string | null {
  const lower = inside.toLowerCase()
  // 找 [0-9]+[mk] 子串
  const m = lower.match(/(\d+)([mk])/)
  return m ? `${m[1]}${m[2]}` : null
}

function compactParts(parts: string[]): string {
  if (parts.length === 0) return ''
  if (parts.length === 1) return parts[0]

  // name 段尾巴
  let splitAt = 0
  for (let i = 0; i < parts.length; i++) {
    if (!isVersionish(parts[i])) splitAt = i + 1
    else break
  }
  if (splitAt === 0 || splitAt >= parts.length) {
    return parts.filter((p) => !isYyyymmdd(p)).join('-')
  }
  const head = parts.slice(0, splitAt).join('-')
  const verTokens: string[] = []
  const suffixTokens: string[] = []
  let inVer = true
  for (const p of parts.slice(splitAt)) {
    if (inVer && isVersionish(p)) verTokens.push(p)
    else {
      inVer = false
      suffixTokens.push(p)
    }
  }
  let out = head
  if (verTokens.length) out += '-' + verTokens.join('.')
  for (const t of suffixTokens) {
    if (!isYyyymmdd(t)) out += '-' + t
  }
  return out
}

function isVersionish(s: string): boolean {
  if (!s || isYyyymmdd(s)) return false
  return /^[\d.]+$/.test(s)
}

function isYyyymmdd(s: string): boolean {
  return s.length === 8 && /^\d{8}$/.test(s)
}

/**
 * 处理 claude code 的"ver 在前、tier 在后"格式。镜像 model_alias.rs::swap_ver_tier_if_needed。
 *   ["4.7", "opus"]            → ["opus", "4.7"]
 *   ["4.7", "opus", "1m"]      → ["opus", "4.7", "1m"]
 *   ["opus", "4.7"]            → 原样
 */
function swapVerTierIfNeeded(parts: string[]): string[] {
  if (parts.length < 2) return parts
  if (!isVersionish(parts[0])) return parts
  let nameIdx = -1
  for (let i = 1; i < parts.length; i++) {
    if (!isVersionish(parts[i])) {
      nameIdx = i
      break
    }
  }
  if (nameIdx === -1) return parts
  return [parts[nameIdx], ...parts.slice(0, nameIdx), ...parts.slice(nameIdx + 1)]
}

/**
 * 模型 → context window(token 数)。镜像 kode-core/src/context.rs。
 * 未知返回 null,UI 应显示 "—"。
 */
export function contextWindow(model: string): number | null {
  const m = (model ?? '').trim().toLowerCase()
  if (!m || m === 'auto') return null

  if (m.includes('1m')) return 1_000_000
  if (m.includes('200k')) return 200_000
  if (m.includes('128k')) return 128_000

  const after = m.replace(/^anthropic\./, '').replace(/^claude-/, '')
  if (after.startsWith('opus') || after.startsWith('sonnet') || after.startsWith('haiku')) {
    return 200_000
  }
  if (after.startsWith('gpt-5') || m.startsWith('gpt-5')) return 400_000
  if (after.startsWith('gpt-4o') || m.startsWith('gpt-4o')) return 128_000
  if (after.startsWith('gpt-4') || m.startsWith('gpt-4')) return 128_000
  if (after.startsWith('gpt-3.5') || m.startsWith('gpt-3.5')) return 16_000
  if (m.startsWith('gemini')) return 1_000_000
  if (
    m.startsWith('glm') ||
    m.startsWith('kimi') ||
    m.startsWith('minimax') ||
    m.startsWith('deepseek') ||
    m.startsWith('hy')
  )
    return 128_000

  return null
}

/** 把 backend_key 映射成展示用的 chip 文案与 CSS class。
 *  tint:用于 compact 模式 tile 背景渐变(空字符串 = 无 tint,用默认色)。 */
export function backendChip(key: string): { label: string; cls: string; tint: string } {
  const k = (key ?? '').toLowerCase()
  if (k === 'codebuddy') return { label: 'codebuddy', cls: 'chip-codebuddy', tint: '#9FE870' }
  if (k === 'claude') return { label: 'claude', cls: 'chip-claude', tint: '#D97757' }
  if (k === 'claude-internal') return { label: 'claude-int', cls: 'chip-claude', tint: '#D97757' }
  if (k.startsWith('claude-')) return { label: `claude-${k.slice('claude-'.length, 'claude-'.length + 3)}`, cls: 'chip-claude', tint: '#D97757' }
  if (k.startsWith('claude')) return { label: 'claude', cls: 'chip-claude', tint: '#D97757' }
  if (k.startsWith('codex')) return { label: k, cls: 'chip-other', tint: '#7A9DFF' }
  if (k.startsWith('gemini')) return { label: k, cls: 'chip-other', tint: '#3186FF' }
  if (k.startsWith('cursor')) return { label: k, cls: 'chip-other', tint: '#F54E00' }
  if (k.startsWith('copilot') || k.startsWith('github-copilot')) return { label: k, cls: 'chip-other', tint: '#6E40C9' }
  if (k.startsWith('kimi')) return { label: k, cls: 'chip-other', tint: '#C9C3D6' }
  if (k.startsWith('kiro')) return { label: k, cls: 'chip-other', tint: '#9046FF' }
  if (k.startsWith('amp')) return { label: k, cls: 'chip-other', tint: '#E8B168' }
  if (k.startsWith('grok')) return { label: k, cls: 'chip-other', tint: '#E8E8E8' }
  if (k.startsWith('opencode')) return { label: k, cls: 'chip-other', tint: '#B0B0B0' }
  if (k.startsWith('droid')) return { label: k, cls: 'chip-other', tint: '#C9CDD3' }
  if (k.startsWith('antigravity') || k.startsWith('agy')) return { label: k, cls: 'chip-other', tint: '#4285F4' }
  if (k.startsWith('pi')) return { label: k, cls: 'chip-other', tint: '#C2C5CE' }
  return { label: k || 'unknown', cls: 'chip-other', tint: '' }
}

/**
 * 超短模型名 —— 给 tab chip 这种宽度极度受限的场景。
 *
 * 基于 shortModelName() 的输出再做一次品牌缩写,
 * 目标 ~5-8 字符。未知模型直接退回到 shortModelName()。
 *
 * 输入示例:
 *   "claude-opus-4.7-1m"           → "Opus 4.7"
 *   "deepseek-v4-pro-ioa"         → "DS v4p"
 *   "kimi-k2.6-ioa"               → "Kimi 2.6"
 *   "hy3-preview-ioa"             → "HY 3"
 *   "gpt-5.3-codex"               → "GPT 5.3c"
 */
export function compactModelName(raw: string): string {
  const s = shortModelName(raw)
  if (!s || s === 'auto') return s

  const lower = s.toLowerCase()

  // --- Claude family ---
  if (lower.startsWith('opus')) return compactClaude('Opus', s)
  if (lower.startsWith('sonnet')) return compactClaude('Sonnet', s)
  if (lower.startsWith('haiku')) return compactClaude('Haiku', s)

  // --- GPT family ---
  if (lower.startsWith('gpt-')) {
    // gpt-5.3-codex → GPT 5.3c
    const rest = lower.slice(4) // "5.3-codex"
    const parts = rest.split('-')
    const ver = parts[0] // "5.3"
    const suffix = parts.length > 1 ? parts[parts.length - 1] : '' // "codex"
    if (suffix && suffix !== ver) {
      return `GPT ${ver}${suffix[0]}`
    }
    return `GPT ${ver}`
  }

  // --- Gemini ---
  if (lower.startsWith('gemini')) {
    // gemini-3.1-pro → Gem 3.1p
    const rest = lower.slice(7).replace(/^[-.]+/, '') // "3.1-pro"
    const parts = rest.split('-')
    const ver = parts[0] // "3.1"
    const suffix = parts.length > 1 ? parts[parts.length - 1] : '' // "pro"
    if (suffix && suffix !== ver) {
      return `Gem ${ver}${suffix[0]}`
    }
    return `Gem ${ver}`
  }

  // --- DeepSeek ---
  if (lower.startsWith('deepseek')) {
    // deepseek-v4-pro-ioa → DS v4p
    const rest = lower.slice(9).replace(/^[-.]+/, '') // "v4-pro-ioa"
    const parts = rest.split('-')
    const ver = parts[0] // "v4"
    // 第二个有意义段(跳过 ver 段)
    let tier = ''
    for (let i = 1; i < parts.length; i++) {
      if (parts[i] && !isVersionish(parts[i])) {
        tier = parts[i]
        break
      }
    }
    if (tier) {
      return `DS ${ver}${tier[0]}`
    }
    return `DS ${ver}`
  }

  // --- Kimi ---
  if (lower.startsWith('kimi')) {
    // kimi-k2.6-ioa → Kimi 2.6
    const rest = lower.slice(5).replace(/^[-.]+/, '') // "k2.6-ioa"
    const verMatch = rest.match(/([kv]?\d[\d.]*)/)
    if (verMatch) {
      const ver = verMatch[0].replace(/^[kv]/, '') // "2.6"
      return `Kimi ${ver}`
    }
    return 'Kimi'
  }

  // --- GLM ---
  if (lower.startsWith('glm')) {
    // glm-4.7-ioa → GLM 4.7
    const rest = lower.slice(4).replace(/^[-.]+/, '')
    const verMatch = rest.match(/(\d[\d.]*)/)
    if (verMatch) {
      return `GLM ${verMatch[0]}`
    }
    return 'GLM'
  }

  // --- MiniMax ---
  if (lower.startsWith('minimax')) {
    // minimax-m2.1 → Mini 2.1
    const rest = lower.slice(8).replace(/^[-.]+/, '')
    const verMatch = rest.match(/(\d[\d.]*)/)
    if (verMatch) {
      return `Mini ${verMatch[0]}`
    }
    return 'Mini'
  }

  // --- HY (Hunyuan) ---
  if (lower.startsWith('hy')) {
    // hy3-preview-ioa → HY 3
    const rest = lower.slice(2).replace(/^[-.]+/, '') // "3-preview-ioa"
    const verMatch = rest.match(/(\d[\d.]*)/)
    if (verMatch) {
      return `HY ${verMatch[0]}`
    }
    return 'HY'
  }

  // Fallback: return shortModelName as-is
  return s
}

/**
 * 极短模型缩写(2-4 字符),给 compact 侧栏 tab tile 底部用。
 * avatar 已表达 backend,这里只取版本号这个互补维度:
 *   glm-5.2-ioa → "5.2"  claude-opus-4.7 → "4.7"  deepseek-v4-pro → "v4"
 *   gpt-5.3-codex → "5.3"  gemini-3.1-pro → "3.1"  kimi-k2.6 → "2.6"
 *   auto / 空 → "auto" / "·"
 */
export function modelAbbr(raw: string | null | undefined): string {
  if (!raw) return '·'
  const s = sanitizeModelName(raw)
  if (!s || s === 'auto') return s || 'auto'
  const lower = s.toLowerCase()

  // deepseek-v4: 保留 v 前缀(版本的一部分)
  if (lower.startsWith('deepseek')) {
    const m = lower.match(/v\d[\d.]*/)
    if (m) return m[0]
  }

  // kimi-k2.6 / kv3 → 去掉 k 前缀取版本
  if (lower.startsWith('kimi')) {
    const m = lower.match(/[kv]?(\d[\d.]*)/)
    if (m) return m[1]
  }

  // 其余:取第一个「纯数字版本段」(含小数点)
  const m = lower.match(/(\d[\d.]*)/)
  return m ? m[1] : '·'
}

/** Claude family compact: "Opus 4.7 1M" / "Opus 4.7" */
function compactClaude(tier: string, s: string): string {
  const parts = s.toLowerCase().split('-')
  // tier 已消费,后面是 ver + optional size
  let ver = ''
  let size = ''
  for (const p of parts) {
    if (!ver && isVersionish(p)) {
      ver = p
      continue
    }
    if (ver && /^\d+[mk]$/i.test(p)) {
      size = p.toUpperCase()
      break
    }
  }
  if (size) return `${tier} ${ver} ${size}`
  if (ver) return `${tier} ${ver}`
  return tier
}

/** 把 token 数压成 12k / 1.2M 形式 */
export function formatTokens(n: number): string {
  if (n >= 1_000_000) return (n / 1_000_000).toFixed(1) + 'M'
  if (n >= 1_000) return (n / 1_000).toFixed(1) + 'k'
  return String(n)
}

File: crates/kode-core/src/model_alias.rs
//! 模型简称归一 —— UI 层(TUI / GUI / 未来手机端)展示模型时的统一短名。
//!
//! 输入:codebuddy / claude / openai / gemini 等任意来源的模型字符串。
//! 输出:适合展示在窄列里的短名(目标 ≤ 14 字符)。
//!
//! 典型映射:
//! ```text
//!   claude-opus-4.7              → opus-4.7
//!   claude-opus-4.7-1m           → opus-4.7-1m
//!   claude-sonnet-4-5-20250929   → sonnet-4.5
//!   "Claude-Sonnet-4.6 (1M context)" → sonnet-4.6-1m
//!   anthropic.claude-haiku-4.5   → haiku-4.5
//!   gpt-5.3-codex                → gpt-5.3-codex
//!   gemini-3.1-pro               → gemini-3.1-pro
//!   ""  / "auto"                 → 原样
//! ```
//!
//! GUI 前端有一份 TS 镜像(`apps/gui/src/lib/model_alias.ts`),双方共享测试夹具
//! `tests/model_alias_fixtures.json`,确保两实现一致。

/// 主入口:把任意来源的模型字符串压成短展示名。
pub fn short_model_name(raw: &str) -> String {
    let cleaned = sanitize_model_name(raw);
    let s = cleaned.trim();
    if s.is_empty() || s == "auto" {
        return s.to_string();
    }

    // 1) 处理"友好名"格式:"Claude-Sonnet-4.6 (1M context)" / "Claude Opus 4.7 (1M)"
    //    思路:抓括号外的主体 + 抓括号内的"1M / 200K"标签作 suffix
    let (body, paren_suffix) = split_paren(s);

    // 2) 主体内部:转小写 + 把空格换成 dash
    let body = body.trim().replace(' ', "-").to_ascii_lowercase();

    // 3) 去公共前缀(anthropic. / claude- / claude-)
    let body = body
        .trim_start_matches("anthropic.")
        .trim_start_matches("claude-");

    // 4) 把以 dash 分隔的版本号 token 合并回 "4.7" 形式
    //    并丢掉 yyyymmdd
    let parts: Vec<&str> = body.split('-').filter(|p| !p.is_empty()).collect();
    // 4.5) claude code 用 "claude-<ver>-<tier>" 顺序(如 "claude-4.7-opus");
    //      codebuddy / Anthropic 公开 ID 用 "claude-<tier>-<ver>" 顺序。
    //      统一成后者再走 compact_parts。
    let parts = swap_ver_tier_if_needed(parts);
    let compact = compact_parts(&parts);

    // 5) 拼上括号内的简化标签(如 "1m"),若主体里没已经带这个标签
    if let Some(tag) = paren_suffix {
        if !compact.contains(&tag) {
            return format!("{compact}-{tag}");
        }
    }
    compact
}

/// 清理 codebuddy 偶发写进模型字段的 note / 提示后缀,保留真正的模型名。
///
/// 已见过的脏值形态:
/// - `Claude-Opus-4.8-1M Note: The model was saved ...`
/// - `opus-4.8-1m-note:-the-model-was-saved-to-user-settings,...`
/// - `glm-5.2-ioa\nNote: The model was saved to user settings, ...`(note 用换行分隔)
pub fn sanitize_model_name(raw: &str) -> String {
    let s = raw.trim();
    if s.is_empty() {
        return String::new();
    }

    let lower = s.to_ascii_lowercase();
    let mut cut = s.len();
    for marker in [
        " note:",
        "-note:",
        "\nnote:",
        "\r\nnote:",
        "\tnote:",
        " the model was saved to user settings",
        "-the-model-was-saved-to-user-settings",
    ] {
        if let Some(idx) = lower.find(marker) {
            cut = cut.min(idx);
        }
    }

    // 兜底:合法 model 名只含 [a-z0-9._-](大小写都行),不会含 \n / \r / \t。
    // 友好名最多用空格(如 "Claude Opus 4.7"),所以这三个字符出现就一定是
    // codebuddy 又夹带了新的 note 形态 —— 直接在第一个 \n/\r/\t 处截断。
    // 这条 fallback 保证未来 codebuddy 换分隔符时不会再漏(只需 marker 列表覆盖空格 case)。
    if let Some(idx) = lower.find(|c: char| matches!(c, '\n' | '\r' | '\t')) {
        cut = cut.min(idx);
    }

    s[..cut]
        .trim_end_matches(|c: char| matches!(c, ' ' | '-'))
        .trim()
        .to_string()
}

/// 把 "Foo (Bar baz)" 拆成 ("Foo", Some("bar"))。
/// 括号内文本会取第一个能识别的 size token(1m / 200k / 128k …),
/// 找不到就返回 None(把括号整段丢了,因为对展示无意义)。
fn split_paren(s: &str) -> (String, Option<String>) {
    if let (Some(lp), Some(rp)) = (s.find('('), s.rfind(')')) {
        if rp > lp {
            let body = format!("{}{}", &s[..lp], &s[rp + 1..]);
            let inside = &s[lp + 1..rp];
            let tag = extract_size_tag(inside);
            return (body.trim().to_string(), tag);
        }
    }
    (s.to_string(), None)
}

/// 从 "1M context" / "200k tokens" 之类的字符串里抠出标准化 size 标签。
/// 返回小写无单位前后缀,如 "1m" / "200k" / "128k"。
fn extract_size_tag(inside: &str) -> Option<String> {
    let lower = inside.to_ascii_lowercase();
    // 找连续 [0-9]+[mk] 的子串
    let bytes = lower.as_bytes();
    let mut i = 0;
    while i < bytes.len() {
        if bytes[i].is_ascii_digit() {
            let start = i;
            while i < bytes.len() && bytes[i].is_ascii_digit() {
                i += 1;
            }
            if i < bytes.len() && (bytes[i] == b'k' || bytes[i] == b'm') {
                let unit = bytes[i];
                i += 1;
                return Some(format!("{}{}", &lower[start..i - 1], unit as char));
            }
        } else {
            i += 1;
        }
    }
    None
}

/// 给定 token 列表(已小写,已 trim),把数字段合并、丢日期,输出展示串。
fn compact_parts(parts: &[&str]) -> String {
    if parts.is_empty() {
        return String::new();
    }
    if parts.len() == 1 {
        return parts[0].to_string();
    }

    // 找 name 段尾巴(最后一个不是纯数字的 token 之前的部分)
    let mut split_at = 0usize;
    for (i, p) in parts.iter().enumerate() {
        if !is_versionish(p) {
            split_at = i + 1;
        } else {
            break;
        }
    }
    if split_at == 0 || split_at >= parts.len() {
        // 没有数字尾巴 → 原样 join,但仍丢掉日期
        return parts
            .iter()
            .filter(|p| !is_yyyymmdd(p))
            .copied()
            .collect::<Vec<_>>()
            .join("-");
    }
    let head = parts[..split_at].join("-");
    let mut ver_tokens = vec![];
    let mut suffix_tokens = vec![];
    let mut in_ver = true;
    for p in &parts[split_at..] {
        if in_ver && is_versionish(p) {
            ver_tokens.push(*p);
        } else {
            in_ver = false;
            suffix_tokens.push(*p);
        }
    }
    let mut out = head;
    if !ver_tokens.is_empty() {
        out.push('-');
        out.push_str(&ver_tokens.join("."));
    }
    for t in suffix_tokens {
        if !is_yyyymmdd(t) {
            out.push('-');
            out.push_str(t);
        }
    }
    out
}

fn is_versionish(s: &str) -> bool {
    if s.is_empty() || is_yyyymmdd(s) {
        return false;
    }
    s.chars().all(|c| c.is_ascii_digit() || c == '.')
}

fn is_yyyymmdd(s: &str) -> bool {
    s.len() == 8 && s.chars().all(|c| c.is_ascii_digit())
}

/// 处理 claude code 的"ver 在前、tier 在后"格式。
///
/// 输入(已 trim 掉 claude- 前缀):
///   ["4.7", "opus"]            → ["opus", "4.7"]
///   ["4.7", "opus", "1m"]      → ["opus", "4.7", "1m"]
///   ["opus", "4.7"]            → ["opus", "4.7"]   (codebuddy 风,不动)
///   ["foo", "bar"]             → ["foo", "bar"]   (没有 versionish 头,不动)
fn swap_ver_tier_if_needed<'a>(parts: Vec<&'a str>) -> Vec<&'a str> {
    if parts.len() < 2 {
        return parts;
    }
    let first_is_ver = is_versionish(parts[0]);
    if !first_is_ver {
        return parts;
    }
    // 找下一个非 versionish 段(name token)的位置
    let name_idx = parts.iter().skip(1).position(|p| !is_versionish(p));
    let Some(name_idx) = name_idx else {
        return parts;
    };
    let name_idx = name_idx + 1; // 因为 skip(1)
                                 // 拆三段:vers_head | name | rest
    let mut out = Vec::with_capacity(parts.len());
    out.push(parts[name_idx]);
    out.extend_from_slice(&parts[..name_idx]);
    out.extend_from_slice(&parts[name_idx + 1..]);
    out
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn passthrough_empty_and_auto() {
        assert_eq!(short_model_name(""), "");
        assert_eq!(short_model_name("auto"), "auto");
        assert_eq!(short_model_name("  "), "");
    }

    // 现有 TUI 测试覆盖的 case
    #[test]
    fn claude_opus_dotted() {
        assert_eq!(short_model_name("claude-opus-4.7"), "opus-4.7");
    }

    #[test]
    fn claude_opus_with_1m_suffix() {
        assert_eq!(short_model_name("claude-opus-4.7-1m"), "opus-4.7-1m");
    }

    #[test]
    fn anthropic_dashed_version_with_date() {
        assert_eq!(short_model_name("claude-sonnet-4-5-20250929"), "sonnet-4.5");
    }

    #[test]
    fn gpt_passthrough() {
        assert_eq!(short_model_name("gpt-5.3-codex"), "gpt-5.3-codex");
    }

    #[test]
    fn gemini_passthrough() {
        assert_eq!(short_model_name("gemini-3.1-pro"), "gemini-3.1-pro");
    }

    #[test]
    fn unknown_raw() {
        assert_eq!(short_model_name("foo"), "foo");
    }

    // 新增:友好名格式
    #[test]
    fn friendly_sonnet_4_6_with_paren() {
        assert_eq!(
            short_model_name("Claude-Sonnet-4.6 (1M context)"),
            "sonnet-4.6-1m"
        );
    }

    #[test]
    fn friendly_opus_with_space_and_paren() {
        assert_eq!(short_model_name("Claude Opus 4.7 (1M)"), "opus-4.7-1m");
    }

    #[test]
    fn friendly_sonnet_4_6_default_200k() {
        // 没括号或括号里没 size token → 不加 suffix
        assert_eq!(short_model_name("Claude-Sonnet-4.6"), "sonnet-4.6");
    }

    #[test]
    fn anthropic_prefix_dotted() {
        assert_eq!(short_model_name("anthropic.claude-haiku-4.5"), "haiku-4.5");
    }

    // 已有 1m 在主体里 → 不重复加
    #[test]
    fn does_not_double_1m() {
        assert_eq!(
            short_model_name("Claude-Sonnet-4.6-1m (1M context)"),
            "sonnet-4.6-1m"
        );
    }

    // 200k 标签
    #[test]
    fn paren_200k_tag() {
        assert_eq!(
            short_model_name("Claude-Sonnet-4.6 (200K context)"),
            "sonnet-4.6-200k"
        );
    }

    // ── claude code 的 ver-tier 顺序(claude-<ver>-<tier>)
    #[test]
    fn claude_code_ver_tier_opus() {
        assert_eq!(short_model_name("claude-4.7-opus"), "opus-4.7");
    }

    #[test]
    fn claude_code_ver_tier_with_1m_suffix() {
        assert_eq!(short_model_name("claude-4.7-opus-1m"), "opus-4.7-1m");
    }

    #[test]
    fn claude_code_ver_tier_sonnet() {
        assert_eq!(short_model_name("claude-4.6-sonnet"), "sonnet-4.6");
    }

    #[test]
    fn claude_code_friendly_name_with_paren() {
        // 实际看到过的形式
        assert_eq!(
            short_model_name("Claude-4.7-Opus (1M context)"),
            "opus-4.7-1m"
        );
    }

    // ── 防御:模型名后跟了 note/警告文本
    #[test]
    fn strips_note_suffix_after_space() {
        assert_eq!(
            short_model_name("Claude-Opus-4.8-1M Note: the model was saved to user settings"),
            "opus-4.8-1m"
        );
    }

    #[test]
    fn strips_hyphenated_note_suffix() {
        assert_eq!(
            short_model_name(
                "opus-4.8-1m-note:-the-model-was-saved-to-user-settings,-but-the-project-\"model\"-setting-will-override-it-after-restart"
            ),
            "opus-4.8-1m"
        );
    }

    // 回归:codebuddy 偶发把带换行符的 note 塞进 requestModelName
    // 真实线上形态:`glm-5.2-ioa\nNote: The model was saved to user settings, ...`
    // 之前的 marker 列表只认 " note:"(空格前导),\n 不匹配 → 整段带换行的脏值
    // 被原样当 model 名注入 `--model` argv,子进程收到后异常。
    // 现在用「任意 ASCII 空白即截断」兜底。
    #[test]
    fn strips_note_suffix_after_newline() {
        assert_eq!(
            sanitize_model_name(
                "glm-5.2-ioa\nNote: The model was saved to user settings, but the project \"model\" setting will override it after restart. Remove the project \"model\" setting to persist this choice."
            ),
            "glm-5.2-ioa"
        );
        // \r\n 也算
        assert_eq!(sanitize_model_name("glm-5.2-ioa\r\nNote: x"), "glm-5.2-ioa");
        // \t 也算
        assert_eq!(sanitize_model_name("glm-5.2-ioa\tNote: x"), "glm-5.2-ioa");
        // 空格 + Note 仍然走老路径(回归保护)
        assert_eq!(
            sanitize_model_name("Claude-Opus-4.8-1M Note: The model was saved"),
            "Claude-Opus-4.8-1M"
        );
    }

    // 友好名 + 括号 size tag 仍然生效
    #[test]
    fn friendly_name_with_paren_still_works() {
        assert_eq!(
            short_model_name("Claude-4.7-Opus (1M context)"),
            "opus-4.7-1m"
        );
    }
}
